File: Shelly.Gtk/Resources/Translations.cs
using System.Runtime.InteropServices;

namespace Shelly.GTK.Resources;

internal static partial class Translations
{
    public const string Domain = "shelly-ui";

    [LibraryImport("libc", StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint bindtextdomain(string domainname, string dirname);

    [LibraryImport("libc", StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint bind_textdomain_codeset(string domainname, string codeset);

    [LibraryImport("libc", StringMarshalling = StringMarshalling.Utf8)]
    // ReSharper disable once InconsistentNaming
    private static partial string setlocale(int category, [MarshalAs(UnmanagedType.LPStr)] string locale);

    // ReSharper disable once InconsistentNaming
    private const int LC_ALL = 6;
    
    [LibraryImport("libc", EntryPoint = "setenv", SetLastError = true, StringMarshalling = StringMarshalling.Utf8)]
    private static partial int setenv(string name, string value, int overwrite);
    
    [LibraryImport("libc", EntryPoint = "unsetenv", SetLastError = true, StringMarshalling = StringMarshalling.Utf8)]
    private static partial int unsetenv(string name);

    internal static void Init(string? culture = null)
    {
        if (string.IsNullOrWhiteSpace(culture))
            unsetenv("LANGUAGE");   
        else
            setenv("LANGUAGE", culture, 1);
        
        setlocale(LC_ALL, "");
        const string localeDir = "/usr/share/locale";
        bindtextdomain(Domain, localeDir);
        bind_textdomain_codeset(Domain, "UTF-8");
    }

    internal static string T(string msgid)
    {
        try
        {
            return GLib.Functions.Dgettext(Domain, msgid);
        }
        catch (DllNotFoundException)
        {
            return msgid;
        }
    }

    internal static string T(string msgid, params object[] args)
    {
        try
        {
            return string.Format(GLib.Functions.Dgettext(Domain, msgid), args);
        }
        catch (DllNotFoundException)
        {
            return string.Format(msgid, args);
        }
    }
}

File: Shelly.Ui.Gtk/src/helpers/translations.zig
const std = @import("std");

pub const domain = "shelly-ui";
pub const locale_dir = "/usr/share/locale";

const LC_ALL: c_int = 6;

extern fn setlocale(category: c_int, locale: [*:0]const u8) ?[*:0]const u8;
extern fn bindtextdomain(domainname: [*:0]const u8, dirname: [*:0]const u8) ?[*:0]const u8;
extern fn bind_textdomain_codeset(domainname: [*:0]const u8, codeset: [*:0]const u8) ?[*:0]const u8;
extern fn textdomain(domainname: [*:0]const u8) ?[*:0]const u8;
extern fn dgettext(domainname: [*:0]const u8, msgid: [*:0]const u8) [*:0]const u8;

/// Initializes the process locale and the gettext domain used by both Zig
/// strings and GtkBuilder resources.
pub fn init() bool {
    const locale_ok = setlocale(LC_ALL, "") != null;
    const directory_ok = bindtextdomain(domain, locale_dir) != null;
    const codeset_ok = bind_textdomain_codeset(domain, "UTF-8") != null;
    const domain_ok = textdomain(domain) != null;
    return locale_ok and directory_ok and codeset_ok and domain_ok;
}

pub fn _(msgid: [:0]const u8) [:0]const u8 {
    return std.mem.span(dgettext(domain, msgid.ptr));
}
